src/mokernels/moinput.jl

```diff
@@ -8,24 +8,15 @@ struct MOInput{T<:AbstractVector} <: AbstractVector{Tuple{Any,Int}}
     out_dim::Integer
 end
 
-Base.length(inp::MOInput) = inp.out_dim * length(inp.x)
-
-Base.size(inp::MOInput, d) = d::Integer == 1 ? inp.out_dim * size(inp.x, 1) : 1 
 Base.size(inp::MOInput) = (inp.out_dim * size(inp.x, 1),)
 
-Base.lastindex(inp::MOInput) = length(inp)
-Base.firstindex(inp::MOInput) = 1
-
 function Base.getindex(inp::MOInput, ind::Integer)
-    if ind > 0
-        out_dim = ind ÷ length(inp.x) + 1
-        ind = ind % length(inp.x)
-        if ind==0 ind = length(inp.x); out_dim-=1 end
-        return (inp.x[ind], out_dim::Int)
-    else
+    @boundscheck if ind <= 0 || ind > length(inp)
```

Comment:
As discussed in https://docs.julialang.org/en/v1/devdocs/boundscheck/, this should just be
```suggestion
    @boundscheck checkbounds(inp, ind)
```
and the function has to be inlined for `@inbounds` to have any effect.

Reply:
In particular the custom error in the next lines should be removed.